File: src/agent_harness/tool_executor.py
"""실제 도구 실행기 - 시뮬레이션 대신 실제로 도구를 실행한다."""

from __future__ import annotations

import fnmatch
import logging
import os
import re
import subprocess
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

# Bash 실행 시 차단할 위험 패턴
_DANGEROUS_PATTERNS = [
    r"\brm\s+-rf\b",
    r"\brm\s+-r\b",
    r"\bmkfs\b",
    r"\bdd\s+if=",
    r"\b:[(][)]\s*[{]",          # fork bomb
    r"\bshutdown\b",
    r"\breboot\b",
    r"\bsudo\b",
    r"\bchmod\s+777\b",
    r"\bcurl\b.*\|\s*bash",       # pipe to bash
    r"\bwget\b.*\|\s*bash",
]


class ToolExecutionError(Exception):
    """도구 실행 중 발생한 오류."""


class ToolExecutor:
    """프로젝트 루트 기준으로 도구를 실제 실행한다.

    지원 도구: Read, Grep, Glob, Bash (제한적)
    안전하지 않은 도구(Write, Edit)는 기본 비활성.
    """

    def __init__(
        self,
        project_root: str | Path,
        allowed_tools: set[str] | None = None,
        bash_timeout: int = 30,
    ):
        self._root = Path(project_root).resolve()
        if not self._root.is_dir():
            raise FileNotFoundError(f"프로젝트 경로가 존재하지 않습니다: {self._root}")

        # 기본: 읽기 전용 도구만 허용
        self._allowed = allowed_tools or {"Read", "Grep", "Glob"}
        self._bash_timeout = bash_timeout

    @property
    def project_root(self) -> Path:
        return self._root

    def _resolve_path(self, file_path: str) -> Path:
        """파일 경로를 프로젝트 루트 기준으로 해석.

        절대 경로면 그대로, 상대 경로면 프로젝트 루트 기준으로 해석.
        프로젝트 루트 바깥 경로는 차단.
        """
        p = Path(file_path)
        if not p.is_absolute():
            p = self._root / p
        p = p.resolve()

        # 프로젝트 루트 바깥 접근 차단
        try:
            p.relative_to(self._root)
        except ValueError:
            raise ToolExecutionError(
                f"프로젝트 루트 바깥 경로 접근 차단: {file_path} → {p}"
            )

        return p

    def execute(self, tool_name: str, tool_input: dict[str, Any]) -> str:
        """도구를 실제로 실행하고 결과를 반환한다.

        허용되지 않은 도구는 시뮬레이션 결과를 반환.
        """
        if tool_name not in self._allowed:
            return f"[harness] {tool_name} 도구는 실제 실행이 허용되지 않음 (simulated)"

        dispatch = {
            "Read": self._execute_read,
            "Grep": self._execute_grep,
            "Glob": self._execute_glob,
            "Bash": self._execute_bash,
        }

        handler = dispatch.get(tool_name)
        if handler is None:
            return f"[harness] {tool_name} 도구는 실행기에 구현되지 않음 (simulated)"

        try:
            return handler(tool_input)
        except ToolExecutionError as e:
            return f"[error] {e}"
        except Exception as e:
            logger.exception("도구 실행 중 예외: %s", tool_name)
            return f"[error] {tool_name} 실행 실패: {e}"

    def _execute_read(self, tool_input: dict[str, Any]) -> str:
        """Read 도구 실행 - 파일 내용을 읽어 반환."""
        file_path = tool_input.get("file_path", "")
        if not file_path:
            return "[error] file_path가 지정되지 않음"

        resolved = self._resolve_path(file_path)
        if not resolved.is_file():
            return f"[error] 파일을 찾을 수 없음: {file_path}"

        offset = tool_input.get("offset", 0)
        limit = tool_input.get("limit", 2000)

        try:
            lines = resolved.read_text(encoding="utf-8").splitlines()
        except UnicodeDecodeError:
            return f"[error] 바이너리 파일은 읽을 수 없음: {file_path}"

        # offset/limit 적용
        selected = lines[offset:offset + limit]
        numbered = [
            f"{i + offset + 1:>6}\t{line}"
            for i, line in enumerate(selected)
        ]
        return "\n".join(numbered)

    def _execute_grep(self, tool_input: dict[str, Any]) -> str:
        """Grep 도구 실행 - 파일 내용을 정규식으로 검색."""
        pattern_str = tool_input.get("pattern", "")
        if not pattern_str:
            return "[error] pattern이 지정되지 않음"

        search_path = tool_input.get("path", str(self._root))
        resolved = self._resolve_path(search_path)

        try:
            regex = re.compile(pattern_str)
        except re.error as e:
            return f"[error] 잘못된 정규식: {e}"

        glob_filter = tool_input.get("glob")
        output_mode = tool_input.get("output_mode", "files_with_matches")
        case_insensitive = tool_input.get("-i", False)
        if case_insensitive:
            regex = re.compile(pattern_str, re.IGNORECASE)

        matches: list[str] = []
        max_results = 100

        files = self._collect_files(resolved, glob_filter)

        for file_path in files:
            if len(matches) >= max_results:
                break
            try:
                content = file_path.read_text(encoding="utf-8")
            except (UnicodeDecodeError, OSError):
                continue

            file_matches = []
            for i, line in enumerate(content.splitlines(), 1):
                if regex.search(line):
                    file_matches.append((i, line))

            if file_matches:
                rel = file_path.relative_to(self._root)
                if output_mode == "files_with_matches":
                    matches.append(str(rel))
                elif output_mode == "content":
                    for line_no, line in file_matches:
                        matches.append(f"{rel}:{line_no}:{line}")
                elif output_mode == "count":
                    matches.append(f"{rel}:{len(file_matches)}")

        if not matches:
            return f"[no matches for pattern: {pattern_str}]"
        return "\n".join(matches)

    def _execute_glob(self, tool_input: dict[str, Any]) -> str:
        """Glob 도구 실행 - 파일 패턴 매칭."""
        pattern = tool_input.get("pattern", "")
        if not pattern:
            return "[error] pattern이 지정되지 않음"

        search_path = tool_input.get("path", str(self._root))
        resolved = self._resolve_path(search_path)

        matched: list[str] = []
        max_results = 200

        for path in sorted(resolved.rglob("*")):
            if len(matched) >= max_results:
                break
            if not path.is_file():
                continue
            rel = str(path.relative_to(self._root))
            if fnmatch.fnmatch(rel, pattern) or fnmatch.fnmatch(path.name, pattern):
                matched.append(rel)

        if not matched:
            return f"[no files matching: {pattern}]"
        return "\n".join(matched)

    def _execute_bash(self, tool_input: dict[str, Any]) -> str:
        """Bash 도구 실행 - 제한된 쉘 명령 실행."""
        command = tool_input.get("command", "")
        if not command:
            return "[error] command가 지정되지 않음"

        # 위험 명령 차단
        for dangerous in _DANGEROUS_PATTERNS:
            if re.search(dangerous, command):
                return f"[blocked] 위험한 명령이 감지됨: {command}"

        try:
            result = subprocess.run(
                command,
                shell=True,
                capture_output=True,
                text=True,
                timeout=self._bash_timeout,
                cwd=str(self._root),
            )
            output = result.stdout
            if result.stderr:
                output += f"\n[stderr]\n{result.stderr}"
            if result.returncode != 0:
                output += f"\n[exit code: {result.returncode}]"
            return output.strip() or "[no output]"
        except subprocess.TimeoutExpired:
            return f"[timeout] 명령이 {self._bash_timeout}초 내에 완료되지 않음"

    def _collect_files(
        self, path: Path, glob_filter: str | None = None
    ) -> list[Path]:
        """경로에서 파일 목록을 수집한다."""
        if path.is_file():
            return [path]

        files: list[Path] = []
        max_files = 1000
        for p in sorted(path.rglob("*")):
            if len(files) >= max_files:
                break
            if not p.is_file():
                continue
            # .git 등 무시
            parts = p.relative_to(self._root).parts
            if any(part.startswith(".") and part != "." for part in parts):
                continue
            if glob_filter and not fnmatch.fnmatch(p.name, glob_filter):
                continue
            files.append(p)
        return files

File: src/executor/tool-executor.ts
import fs from 'node:fs';
import path from 'node:path';
import { execSync } from 'node:child_process';
import type { IToolExecutor } from './types.js';

const DANGEROUS_PATTERNS = [
  /\brm\s+-rf\b/,
  /\brm\s+-r\b/,
  /\bmkfs\b/,
  /\bdd\s+if=/,
  /\b:[(][)]\s*[{]/,           // fork bomb
  /\bshutdown\b/,
  /\breboot\b/,
  /\bsudo\b/,
  /\bchmod\s+777\b/,
  /\bcurl\b.*\|\s*bash/,
  /\bwget\b.*\|\s*bash/,
];

export class ToolExecutionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ToolExecutionError';
  }
}

export class ToolExecutor implements IToolExecutor {
  private readonly root: string;
  private readonly allowed: Set<string>;
  private readonly bashTimeout: number;

  constructor(
    projectRoot: string,
    options?: { allowedTools?: Set<string>; bashTimeout?: number },
  ) {
    this.root = path.resolve(projectRoot);
    if (!fs.existsSync(this.root) || !fs.statSync(this.root).isDirectory()) {
      throw new Error(`프로젝트 경로가 존재하지 않습니다: ${this.root}`);
    }
    this.allowed = options?.allowedTools ?? new Set(['Read', 'Grep', 'Glob']);
    this.bashTimeout = options?.bashTimeout ?? 30;
  }

  get projectRoot(): string {
    return this.root;
  }

  private resolvePath(filePath: string): string {
    let p = path.isAbsolute(filePath) ? filePath : path.join(this.root, filePath);
    p = path.resolve(p);

    if (!p.startsWith(this.root)) {
      throw new ToolExecutionError(
        `프로젝트 루트 바깥 경로 접근 차단: ${filePath} → ${p}`,
      );
    }
    return p;
  }

  execute(toolName: string, toolInput: Record<string, unknown>): string {
    if (!this.allowed.has(toolName)) {
      return `[harness] ${toolName} 도구는 실제 실행이 허용되지 않음 (simulated)`;
    }

    const dispatch: Record<string, (input: Record<string, unknown>) => string> = {
      Read: (input) => this.executeRead(input),
      Grep: (input) => this.executeGrep(input),
      Glob: (input) => this.executeGlob(input),
      Bash: (input) => this.executeBash(input),
    };

    const handler = dispatch[toolName];
    if (!handler) {
      return `[harness] ${toolName} 도구는 실행기에 구현되지 않음 (simulated)`;
    }

    try {
      return handler(toolInput);
    } catch (e) {
      if (e instanceof ToolExecutionError) {
        return `[error] ${e.message}`;
      }
      return `[error] ${toolName} 실행 실패: ${e}`;
    }
  }

  private executeRead(toolInput: Record<string, unknown>): string {
    const filePath = (toolInput.file_path as string) ?? '';
    if (!filePath) return '[error] file_path가 지정되지 않음';

    const resolved = this.resolvePath(filePath);
    try {
      if (!fs.statSync(resolved).isFile()) {
        return `[error] 파일을 찾을 수 없음: ${filePath}`;
      }
    } catch {
      return `[error] 파일을 찾을 수 없음: ${filePath}`;
    }

    const offset = (toolInput.offset as number) ?? 0;
    const limit = (toolInput.limit as number) ?? 2000;

    let content: string;
    try {
      content = fs.readFileSync(resolved, 'utf-8');
    } catch {
      return `[error] 바이너리 파일은 읽을 수 없음: ${filePath}`;
    }

    const lines = content.split('\n');
    // Remove trailing empty string from split if file ends with newline
    if (lines[lines.length - 1] === '') lines.pop();
    const selected = lines.slice(offset, offset + limit);
    const numbered = selected.map(
      (line, i) => `${String(i + offset + 1).padStart(6)}\t${line}`,
    );
    return numbered.join('\n');
  }

  private executeGrep(toolInput: Record<string, unknown>): string {
    const patternStr = (toolInput.pattern as string) ?? '';
    if (!patternStr) return '[error] pattern이 지정되지 않음';

    const searchPath = (toolInput.path as string) ?? this.root;
    const resolved = this.resolvePath(searchPath);

    let regex: RegExp;
    try {
      const flags = toolInput['-i'] ? 'i' : '';
      regex = new RegExp(patternStr, flags);
    } catch (e) {
      return `[error] 잘못된 정규식: ${e}`;
    }

    const globFilter = toolInput.glob as string | undefined;
    const outputMode = (toolInput.output_mode as string) ?? 'files_with_matches';

    const matches: string[] = [];
    const maxResults = 100;

    const files = this.collectFiles(resolved, globFilter);

    for (const file of files) {
      if (matches.length >= maxResults) break;

      let content: string;
      try {
        content = fs.readFileSync(file, 'utf-8');
      } catch {
        continue;
      }

      const fileMatches: [number, string][] = [];
      const contentLines = content.split('\n');
      for (let i = 0; i < contentLines.length; i++) {
        if (regex.test(contentLines[i])) {
          fileMatches.push([i + 1, contentLines[i]]);
        }
      }

      if (fileMatches.length > 0) {
        const rel = path.relative(this.root, file);
        if (outputMode === 'files_with_matches') {
          matches.push(rel);
        } else if (outputMode === 'content') {
          for (const [lineNo, line] of fileMatches) {
            matches.push(`${rel}:${lineNo}:${line}`);
          }
        } else if (outputMode === 'count') {
          matches.push(`${rel}:${fileMatches.length}`);
        }
      }
    }

    if (matches.length === 0) {
      return `[no matches for pattern: ${patternStr}]`;
    }
    return matches.join('\n');
  }

  private executeGlob(toolInput: Record<string, unknown>): string {
    const pattern = (toolInput.pattern as string) ?? '';
    if (!pattern) return '[error] pattern이 지정되지 않음';

    const searchPath = (toolInput.path as string) ?? this.root;
    const resolved = this.resolvePath(searchPath);

    const matched: string[] = [];
    const maxResults = 200;

    const allFiles = this.collectAllFiles(resolved);
    for (const file of allFiles) {
      if (matched.length >= maxResults) break;
      const rel = path.relative(this.root, file);
      const basename = path.basename(file);
      if (this.matchGlob(rel, pattern) || this.matchGlob(basename, pattern)) {
        matched.push(rel);
      }
    }

    if (matched.length === 0) {
      return `[no files matching: ${pattern}]`;
    }
    return matched.join('\n');
  }

  private executeBash(toolInput: Record<string, unknown>): string {
    const command = (toolInput.command as string) ?? '';
    if (!command) return '[error] command가 지정되지 않음';

    for (const dangerous of DANGEROUS_PATTERNS) {
      if (dangerous.test(command)) {
        return `[blocked] 위험한 명령이 감지됨: ${command}`;
      }
    }

    try {
      const result = execSync(command, {
        cwd: this.root,
        timeout: this.bashTimeout * 1000,
        encoding: 'utf-8',
        stdio: ['pipe', 'pipe', 'pipe'],
      });
      return result.trim() || '[no output]';
    } catch (e: unknown) {
      const err = e as { killed?: boolean; stdout?: string; stderr?: string; status?: number };
      if (err.killed) {
        return `[timeout] 명령이 ${this.bashTimeout}초 내에 완료되지 않음`;
      }
      let output = err.stdout ?? '';
      if (err.stderr) output += `\n[stderr]\n${err.stderr}`;
      if (err.status != null) output += `\n[exit code: ${err.status}]`;
      return output.trim() || '[no output]';
    }
  }

  private collectFiles(dirPath: string, globFilter?: string): string[] {
    const stat = fs.statSync(dirPath);
    if (stat.isFile()) return [dirPath];

    const files: string[] = [];
    const maxFiles = 1000;

    const walk = (dir: string): void => {
      if (files.length >= maxFiles) return;
      let entries: fs.Dirent[];
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }
      entries.sort((a, b) => a.name.localeCompare(b.name));

      for (const entry of entries) {
        if (files.length >= maxFiles) return;
        const fullPath = path.join(dir, entry.name);

        if (entry.isDirectory()) {
          // Skip hidden directories
          if (entry.name.startsWith('.') && entry.name !== '.') continue;
          walk(fullPath);
        } else if (entry.isFile()) {
          const rel = path.relative(this.root, fullPath);
          const parts = rel.split(path.sep);
          if (parts.some((p) => p.startsWith('.') && p !== '.')) continue;
          if (globFilter && !this.matchGlob(entry.name, globFilter)) continue;
          files.push(fullPath);
        }
      }
    };

    walk(dirPath);
    return files;
  }

  private collectAllFiles(dirPath: string): string[] {
    const files: string[] = [];
    const maxFiles = 200;

    const walk = (dir: string): void => {
      if (files.length >= maxFiles) return;
      let entries: fs.Dirent[];
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }
      entries.sort((a, b) => a.name.localeCompare(b.name));

      for (const entry of entries) {
        if (files.length >= maxFiles) return;
        const fullPath = path.join(dir, entry.name);

        if (entry.isDirectory()) {
          walk(fullPath);
        } else if (entry.isFile()) {
          files.push(fullPath);
        }
      }
    };

    walk(dirPath);
    return files;
  }

  private matchGlob(str: string, pattern: string): boolean {
    // Convert glob pattern to regex
    const escaped = pattern
      .replace(/[.+^${}()|[\]\\]/g, '\\$&')
      .replace(/\*\*/g, '<<<GLOBSTAR>>>')
      .replace(/\*/g, '[^/]*')
      .replace(/\?/g, '[^/]')
      .replace(/<<<GLOBSTAR>>>/g, '.*');
    return new RegExp(`^${escaped}$`).test(str);
  }
}
